File: ruby/spec/03_stacks_and_queues/01_three_in_one.rb
# n - one stack size
class NStacksArray
  class StackOverflowError < StandardError
    def initialize(stack)
      super("Stack ##{stack} is overflown")
    end
  end

  attr_reader :stack_size, :stacks_count

  def initialize(stack_size:, stacks_count:)
    @stack_size = stack_size
    @arr = Array.new(stacks_count * stack_size)

    # Indexes of head and tail of stack
    @heads = Array.new(stacks_count)
    @tails = Array.new(stacks_count)

    stacks_count.times do |i|
      @heads[i] = @tails[i] = i * stack_size
    end
  end

  def size(stack)
    si = stack - 1
    @tails[si] - @heads[si]
  end

  def push(stack, value)
    if size(stack) < stack_size
      si = stack - 1
      index = (@tails[si] += 1)
      @arr[index] = value
    else
      raise StackOverflowError.new(stack)
    end
  end

  def pop(stack)
    if size(stack) > 0
      si = stack - 1
      i = @tails[si]
      @tails[si] -= 1
      @arr[i]
    end
  end

  def peek(stack)
    if size(stack) > 0
      @arr[stack - 1]
    end
  end
end

RSpec.describe NStacksArray do
  let(:stack) { described_class.new(stack_size: 5, stacks_count: 3) }

  describe 'push - pop' do
    it do
      stack.push(1, 11)

      stack.push(2, 12)
      stack.push(2, 13)

      stack.push(3, 14)
      stack.push(3, 15)
      stack.push(3, 16)

      expect(stack.pop(1)).to eq 11
      expect(stack.pop(1)).to eq nil

      expect(stack.pop(2)).to eq 13
      expect(stack.pop(2)).to eq 12
      expect(stack.pop(2)).to eq nil

      expect(stack.pop(3)).to eq 16
      expect(stack.pop(3)).to eq 15
      expect(stack.pop(3)).to eq 14
      expect(stack.pop(3)).to eq nil
    end
  end

  describe 'size overflow' do
    it do
      5.times { |i| stack.push(1, i) }
      expect(stack.size(1)).to eq 5

      expect { stack.push(1, 6) }.to raise_error(described_class::StackOverflowError)
    end
  end
end

File: go/03_stacks_and_queues/01_three_in_one.go
package stacks_and_queues

import (
	"errors"
	"fmt"
)

// n - one stack size
type nStacksArray struct {
	stack_size   int
	stacks_count int
	arr          []int
	// Indexes of head and tail of stack
	heads []int
	tails []int
}

func NewNStacksArray(stack_size int, stacks_count int) *nStacksArray {
	sa := &nStacksArray{
		stack_size:   stack_size,
		stacks_count: stacks_count,
		arr:          make([]int, stacks_count*stack_size),
		heads:        make([]int, stacks_count),
		tails:        make([]int, stacks_count),
	}

	for i := 0; i < stacks_count; i++ {
		sa.heads[i] = i * stack_size
		sa.tails[i] = sa.heads[i]
	}

	return sa
}

func (sa *nStacksArray) Size(stack int) int {
	si := stack - 1
	return sa.tails[si] - sa.heads[si]
}

func (sa *nStacksArray) Push(stack int, value int) (bool, error) {
	if sa.Size(stack) < sa.stack_size {
		si := stack - 1
		sa.tails[si] += 1
		index := sa.tails[si]
		sa.arr[index] = value
		return true, nil
	} else {
		return false, errors.New(fmt.Sprintf("Stack #%d is overflown", stack))
	}
}

func (sa *nStacksArray) Pop(stack int) (int, error) {
	if sa.Size(stack) > 0 {
		si := stack - 1
		i := sa.tails[si]
		sa.tails[si] -= 1
		return sa.arr[i], nil
	} else {
		return -1, errors.New(fmt.Sprintf("Stack #%d is empty", stack))
	}
}

func (sa *nStacksArray) Peek(stack int) (int, error) {
	if sa.Size(stack) > 0 {
		return sa.arr[stack-1], nil
	} else {
		return -1, errors.New(fmt.Sprintf("Stack #%d is empty", stack))
	}
}
